### PR TITLE
Roughing in MSI support

### PR DESCRIPTION
**Pull request by directionless**
Initial framework for MSI support

Still being developed, won't really work yet

### CODE REVIEW

File: pkg/packagekit/internal/assets.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
+// Code generated by go-bindata.
```
Comment (groob) [inline pkg/packagekit/internal/assets.go]: don't commit the bindata files, add a `generate` step.

File: cmd/package-builder/package-builder.go [hunk 2]
```diff
@@ -295,64 +296,35 @@ func main() {
 	}
 }
 
+// defaultTargets sets the default target based on platform
+func defaultTargets() string {
+	switch runtime.GOOS {
+	case "windows":
+		return "windows-none-msi"
+	case "linux":
+		return "linux-upstart-rpm,linux-upstart-deb"
```
Comment (groob) [inline cmd/package-builder/package-builder.go]: systemd?

File: pkg/packagekit/wix/wix.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+// Candle invokes wix's candle command. This is the wix compiler, It
+// preprocesses and compiles WiX source files into object files
+// (.wixobj).
+func (wix *Wix) Candle(ctx context.Context) error {
+	_, err := wix.execOut(ctx,
+		filepath.Join(wix.wixPath, "candle.exe"),
+		"-nologo",
+		"-arch", wix.msArch,
+		//"-dGoVersion="+version,
```
Comment (groob) [inline pkg/packagekit/wix/wix.go]: remove or use these commented lines.

File: pkg/packagekit/internal/assets/installer.wxs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	Compressed="yes"
+	InstallScope="perMachine"
+	Languages="1033" />
+
+    <!-- This holds the files generated by heat -->
+    <Media Id='1' Cabinet="go.cab" EmbedCab="yes" CompressionLevel="high" />
+
+    <MajorUpgrade AllowDowngrades="yes" />
+
+    <!-- Set INSTALLDIR to "c:\Program Files\Kolide". With appropriate
```
[the remark this answers is not in the document]
Comment (directionless) [reply pkg/packagekit/internal/assets/installer.wxs]: I started roughing in a split between Program Files and ProgramData, but we can just dump everything in ProgramData for now.

File: pkg/packagekit/package_wix.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+	logger := ctxlog.FromContext(ctx)
+
+	if err := isDirectory(po.Root); err != nil {
+		return err
+	}
+
+	buildDir, err := ioutil.TempDir("", "wix-build")
+	if err != nil {
+		panic(err)
```
Comment (groob) [inline pkg/packagekit/package_wix.go]: return

File: pkg/packagekit/package_wix.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	}
+	level.Debug(logger).Log("builddir", buildDir)
+	// TODO remove this on cleanup
+
+	wixTool, err := wix.New(po.Root, buildDir)
+	if err != nil {
+		panic(err)
```
Comment (groob) [inline pkg/packagekit/package_wix.go]: return

File: pkg/packagekit/package_wix.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	if err := isDirectory(po.Root); err != nil {
+		return err
+	}
+
+	buildDir, err := ioutil.TempDir("", "wix-build")
+	if err != nil {
+		panic(err)
+	}
+	level.Debug(logger).Log("builddir", buildDir)
+	// TODO remove this on cleanup
```
Comment (groob) [inline pkg/packagekit/package_wix.go]: todo?

File: pkg/packagekit/wix/doc.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  3. Use `heat` to harvest the file list from the packageRoot
+  4. Use `candle` to take the wxs from (1) and (3) and make a wixobj
+  5. Use `light` to compile thje wixobj into an msi
+
+
+While this is a somewhat agnostic wrapper, it does make several
+assumptions about the underlying process. It is not meant as a
+complete wix wrapper.
+
+So Many Guids
```
Comment (groob) [inline pkg/packagekit/wix/doc.go]: what did you decide on?

File: pkg/packagekit/wix/wix.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+	case "amd64":
+		wix.msArch = "x64"
+	default:
+		return nil, errors.Errorf("unknown arch for windows %s", runtime.GOARCH)
+	}
+
+	return wix, nil
+}
+
+// InstallWXS takes an io.Reader and drops it's content into the builddir as installer.wxs
```
Comment (groob) [inline pkg/packagekit/wix/wix.go]: outdated

File: pkg/packagekit/wix/wix.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+func (wix *Wix) InstallWXS(installWXS []byte) error {
```
Comment (groob) [inline pkg/packagekit/wix/wix.go]: this function doesn't add anything to the api, all it does is call WriteFile. Could just be a oneliner without the helper.

File: pkg/packagekit/wix/wix.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+		"-dcl:high", // compression level
+		"AppFiles.wixobj",
+		"Installer.wixobj",
+		"-out", "out.msi",
+	)
+	if err != nil {
+		return err
+	}
+
+	msiFH, err := os.Open(filepath.Join(wix.buildDir, "out.msi"))
```
Comment (groob) [inline pkg/packagekit/wix/wix.go]: can you os.rename instead?

File: pkg/packagekit/package_wix.go [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/kolide/launcher/pkg/contexts/ctxlog"
+	"github.com/kolide/launcher/pkg/packagekit/internal"
+	"github.com/kolide/launcher/pkg/packagekit/wix"
+	"github.com/pkg/errors"
+	"go.opencensus.io/trace"
+)
+
+//go:generate go-bindata -pkg internal -o internal/assets.go internal/assets/
+
+func PackageWixMSI(ctx context.Context, w io.Writer, po *PackageOptions) error {
```
Comment (groob) [inline pkg/packagekit/package_wix.go]: why is the output writer not part of the package options?

**Pull request merged**